data/perl/winding.3
.TH WINDING 3 "$Date$"
.SH NAME
winding.pl \- winding number routines for Perl
.SH SYNOPSIS
.nf
require "winding.pl";

get_containing_regions( @db, $lat, $lon );
is_inside_region(  @db, $lat, $lon [, $region_name] );
get_region_polygon( @db, $region_name );
dbview_to_polygon( @db );
is_geographically_inside( $lat, $lon, @polygon );
is_inside_polygon( $x, $y, @polygon );
winding_number( @polygon );
signed_crossing_number( $x1, $y1, $x2, $y2 );
.fi
.SH DESCRIPTION
These routines implement the winding-number algorithm of
Godkin and Pulli, BSSA v.74 1845-1948, 1984. They allow one to take a 
polygon specified by a set of ordered vertices traversed clockwise around 
the polygon, and decide if a specified point is inside or outside the polygon. 

The main routine is is_inside_region(). is_inside_region() takes a
database pointer to a 'regions' table, with fields 
"regname" for unique region names and "vertex" for the order of the defining 
polygon's vertices. See for example the places1.2 schema. If a region_name 
string is specified, is_inside_region looks it up in the table provided.
Otherwise, is_inside_region assumes that the subsetting has already been 
accomplished and the input database view contains only one region. 
is_inside_region returns 0 if the specified point is outside the named region, 
1 if the point is inside. test points on vertices and borders of the region 
will be treated as if they were inside the region. 

The get_containing_regions() function takes a database pointer to a regions 
table and finds the regions that include the specified test point. This 
function returns an array of the names of the regions, or an empty array
if none are found that contain the point. 

In ascending calling order and therefore ascending order of importance to the 
end-user, the other routines are:

signed_crossing_number() takes a line-segment and computes its signed 
crossing number. For explanation please see the referenced article. 

winding_number() takes a polygon, specified as a list of x,y pairs for the 
vertices traversed clockwise and in order, and returns the winding-number of 
the polygon. 

is_inside_polygon() shifts the given polygon to a coordinate system with 
the specified point as the origin, then computes the winding number of the 
polygon to decide if the specified point is inside or outside. 

is_geographically_inside() takes a set of lat,lon coordinates on the surface of 
the earth, transforms to a cartesian coordinate system with a stereographic
projection and calls is_inside_polygon to decide if the reference point 
is inside the polygon. Currently this only works if the reference point and 
the polygon are entirely inside the same hemisphere. 

dbview_to_polygon() takes a view into a regions table, containing rows for 
just one region, and turns it into an array ('polygon') appropriate for use with
the lower level routines. 

get_region_polygon() takes a database pointer containing a regions table, and 
returns the array holding the coordinates of its polygon vertices.
.SH OPTIONS
.SH FILES
.SH ENVIRONMENT
.SH PARAMETER FILE
.SH EXAMPLE
.ft CW
.RS .2i
.nf
@dbregions = dbopen( $region_phrases_database, 'r' );
@dbregions = dblookup( @dbregions, "", "regions", "", "" );

@regions = get_containing_regions( @dbregions, $lat, $lon );

if( ! defined( $where = shift( @regions ) ) ) {

	$where = "(" . grname( $lat, $lon ) . ")";
}
.fi
.RE
.ft R
.SH RETURN VALUES
.SH LIBRARY
.SH DIAGNOSTICS
.SH "SEE ALSO"
.nf
perldb(1)
.fi
.SH "BUGS AND CAVEATS"
A number of internal routines are probably not useful outside, and 
therefore are undocumented. There are a couple test routines which 
demonstrate a number of simple operations with the internal routines. 

The routine is_geographically_inside currently requires that the test 
point and the polygon all lie entirely within the same hemisphere. 
.SH AUTHOR
Algorithm by Godkin and Pulli, BSSA v.74 1845-1948, 1984.
Coding: Kent Lindquist, Geophysical Institute, U. of Alaska. 
.\" $Id$


data/perl/winding/winding.pl
#Perl Reimplementation of winding-number algorithm
#
# Godkin and Pulli, BSSA v.74 1845-1948, 1984. 
#
# K. Lindquist
# Geophysical Institute
# University of Alaska
# 2001

sub get_containing_regions {
	my( @db ) = splice( @_, 0, 4 );
	my( $lat, $lon ) = splice( @_, 0, 2 );
	my( @regnames );
	
	@db = dbsort( @db, "regname" );
	@db = dbgroup( @db, "regname" );

	my( $nregions ) = dbquery( @db, "dbRECORD_COUNT" );

	for( $db[3] = 0; $db[3] < $nregions; $db[3]++ ) {

		my( $regname ) = dbgetv( @db, "regname" );
		my( @polygon ) = get_region_polygon( @db, $regname );

		if( is_geographically_inside( $lat, $lon, @polygon ) ) {

			push( @regnames, $regname );
		}
	}

	return @regnames;
}

sub is_inside_region {
	my( @db ) = splice( @_, 0, 4 );
	my( $lat, $lon ) = splice( @_, 0, 2 );

	if( defined( $region_name = shift( @_ ) ) ) {

		@db = dbsubset( @db, "regname == \"$region_name\"" );

		my( $nrecs ) = dbquery( @db, "dbRECORD_COUNT" );

		if( $nrecs <= 0 ) {
			print STDERR "is_inside_region(): Didn't find " .
				"\"$region_name\" in database.\n";
			return 0;
		}
	}

	my( @polygon ) = dbview_to_polygon( @db );

	return is_geographically_inside( $lat, $lon, @polygon );
}

sub get_region_polygon {
	my( $region_name ) = pop( @_ );
	my( @db ) = @_;

	@db = dblookup( @db, "", "regions", "", "" );
	@db = dbsubset( @db, "regname == \"$region_name\"" );

	return dbview_to_polygon( @db );
}

sub dbview_to_polygon {
	my( @db ) = @_;
	my( $nrecs, @polygon, $lat, $lon, $vertex );
	my( $previous_vertex ) = 0;

	@db = dbsort( @db, "regname", "vertex" );

	$nrecs = dbquery( @db, "dbRECORD_COUNT" );

	if( $nrecs < 1 ) {
		print STDERR 	
			"dbview_to_polygon(): couldn't find region in view\n";
		return ();

	} elsif( $nrecs < 3 ) {
		print STDERR 	
			"dbview_to_polygon(): not enough vertices in view\n";
		return ();
	}

	for( $db[3] = 0; $db[3] < $nrecs; $db[3]++ ) {
		( $lat, $lon, $vertex ) = dbgetv( @db, "lat", "lon", "vertex" );

		if( $vertex < $previous_vertex ) {
			print STDERR "dbview_to_polygon(): More than one " .
				"region in view. Bye.\n";
			return ();
		} else {
			$previous_vertex = $vertex;
		}

		push( @polygon, $lat, $lon );
	}

	return @polygon;
}

# This geographic application of the winding routine is valid
# when the polygon and the test point are entirely within
# the same hemisphere.
sub is_geographically_inside {
	my( $lat, $lon, @polygon ) = @_;
	my( $npoints, $dist, $az, $pi,
	    $point_lat, $point_lon, @geo_polygon );

	if( ! polygon_ok( @polygon ) ) {
		print STDERR "Polygon is Not OK";
		return 0;
	}

	$npoints = ( $#polygon + 1 ) / 2;

	$pi = atan2( 1, 1 ) * 4;

	# polar stereographic projection
	for( $pointnum = 1; $pointnum <= $npoints; $pointnum++ ) {
		$point_lat = $polygon[($pointnum-1)*2];
		$point_lon = $polygon[($pointnum-1)*2+1];

		$dist = dbex_eval( dbinvalid(), 
				"distance($lat,$lon,$point_lat,$point_lon)" );
		$az = dbex_eval( dbinvalid(), 
				"azimuth($lat,$lon,$point_lat,$point_lon)" );

		if( $dist > 90. ) {
			print STDERR "Warning: The routine " .
				"is_geographically_inside() " .
				"is regional only. Polygon and test point " .
				"must be entirely within the same " .
				"hemisphere. Returning FALSE.\n";
			return 0;
		}

		push( @geo_polygon, $dist * sin( $az * $pi/180. ) );
		push( @geo_polygon, $dist * cos( $az * $pi/180. ) );
	}

	return is_inside_polygon( 0, 0, @geo_polygon );
}

sub is_inside_polygon { 
	my( $x, $y, @polygon ) = @_;

	if( ! polygon_ok( @polygon ) ) {
		print STDERR "Polygon is Not OK";
		return 0;
	}

	my( @shifted_polygon ) = shift_polygon( $x, $y, @polygon );

	my( $winding_number ) = winding_number( @shifted_polygon );

	if( $winding_number ) {
		return 1;
	} else {
		return 0;
	}
}

sub shift_polygon {
	my( $x, $y, @polygon ) = @_;
	my( $nsegs, $ix, $iy, $shifted_x, $shifted_y, @shifted_polygon );
	
	$nsegs = ( $#polygon + 1 ) / 2;

	for( $seg = 1; $seg <= $nsegs; $seg++ ) {

		$ix = 2 * ( $seg - 1 );
		$iy = $ix + 1;
		
		$shifted_x = $polygon[$ix] - $x;
		$shifted_y = $polygon[$iy] - $y;
	
		push( @shifted_polygon, $shifted_x );
		push( @shifted_polygon, $shifted_y );
	}

	return @shifted_polygon;
}

sub winding_number {
	my( @shifted_polygon ) = @_;
	my( $ncoords, $nsegs, $winding_number, $seg, @segment, $scn );
	
	# polygon coordinate list must be x1,y1,x2,y2,x3,y3, etc
	
	$ncoords = $#shifted_polygon + 1;
	$nsegs = $ncoords / 2;

	$winding_number = 0;

	for( $seg = 1; $seg <= $nsegs; $seg++ ) {
		
		@segment = segment_number( $seg, @shifted_polygon );
		$scn = signed_crossing_number( @segment );

		$winding_number = $winding_number + $scn;
	}

	return $winding_number;

}

sub polygon_ok {
	my( @polygon ) = @_;

	$ncoords = $#polygon + 1;

	if( $ncoords < 6 ) {
		
		return 0;

	} elsif( is_odd( $ncoords ) ) {
		
		return 0;

	} else {

		return 1;
	}
}

sub is_odd {
	my( $n ) = @_;

	return int( $n / 2 ) != $n / 2;
}

sub segment_number {
	my( $segnum, @polygon ) = @_;
	my( $ncoords, $nsegs, @segment );
	
	$ncoords = $#polygon + 1;

	$nsegs = $ncoords / 2;

	if( $segnum > $nsegs ) {
		print STDERR "ERROR: not that many segments in polygon";
		return ();
	}

	if( $segnum == $nsegs ) {
		push( @segment, $polygon[$ncoords-2] );
		push( @segment, $polygon[$ncoords-1] );
		push( @segment, $polygon[0] );
		push( @segment, $polygon[1] );
	} else {
		push( @segment, $polygon[($segnum-1) * 2] );
		push( @segment, $polygon[($segnum-1) * 2 + 1] );
		push( @segment, $polygon[($segnum-1) * 2 + 2] );
		push( @segment, $polygon[($segnum-1) * 2 + 3] );
	}

	return @segment;
}

sub signed_crossing_number {
	my( $x1, $y1, $x2, $y2 ) = @_;
	my( $xintercept, $slope, $direction );

	if( $y1 * $y2 > 0 ) {
		# no crossing -- both points are on same side of y axis
		return 0;
	}
	
	# Horizontal lines:
	if( $y1 == 0 && $y2 == 0 ) {

		if( $x1 * $x2 > 0 ) {
			# no crossing
			return 0;
		} else {
			# segment crosses or touches origin
			return 2;
		}
	}

	if( $x2 == $x1 ) {
		# Vertical lines
		$xintercept = $x1;
	} else {
		$slope = ( $y2 - $y1 ) / ( $x2 - $x1 );
		$xintercept = $x1 - $y1 / $slope;
	}
	
	if( $y2 > $y1 ) {
		$direction = 1;
	} else {
		$direction = -1;
	}

	if( $xintercept > 0 ) { 

		return 0;

	} elsif( $xintercept == 0 ) {

		return 2;

	} elsif( $y1 == 0 || $y2 == 0 ) {

		return $direction * 0.5;

	} else { 

		return $direction;
	}
}

sub test_winding_signed_crossing_number {
	print STDERR "Above x axis (expect 0): " . 
			signed_crossing_number( 1, 3, 2, 4 ), "\n";
	print STDERR "Below x axis (expect 0): " . 
			signed_crossing_number( 1, -3, 2, -4 ), "\n";
	print STDERR "\n";
	
	print STDERR "Vertical above x axis (expect 0): " . 
			signed_crossing_number( 1, 3, 1, 4 ), "\n";
	print STDERR "On y axis above x axis (expect 0): " . 
			signed_crossing_number( 0, 3, 0, 4 ), "\n";
	print STDERR "On y axis below x axis (expect 0): " . 
			signed_crossing_number( 0, -3, 0, -4 ), "\n";
	print STDERR "Vertical through origin (expect 2): " . 
			signed_crossing_number( 0, -1, 0, 1 ), "\n";
	print STDERR "\n";

	print STDERR "Horizontal on positive x axis (expect 0): " . 
			signed_crossing_number( 2, 0, 4, 0 ), "\n";
	print STDERR "Horizontal on negative x axis (expect 0): " . 
			signed_crossing_number( -2, 0, -4, 0 ), "\n";
	print STDERR "Horizontal through origin (expect 2): " . 
			signed_crossing_number( -1, 0, 1, 0 ), "\n";
	print STDERR "\n";

	print STDERR "starts on origin (expect 2): " . 
			signed_crossing_number( 0, 0, -1, 1 ), "\n";
	print STDERR "ends on origin (expect 2): " . 
			signed_crossing_number( -2, -2, 0, 0 ), "\n";
	print STDERR "\n";

	print STDERR "vertical through positive x axis (expect 0): " . 
			signed_crossing_number( 1, -2, 1, 2 ), "\n";
	print STDERR "oblique through positive x axis (expect 0): " . 
			signed_crossing_number( 1, -2, 2, 2 ), "\n";
	print STDERR "\n";

	print STDERR "vertical up through negative x axis (expect 1): " .
			signed_crossing_number( -1, -2, -1, 2 ), "\n";
	print STDERR "vertical down through negative x axis (expect -1): " .
			signed_crossing_number( -1, 2, -1, -2 ), "\n";
	print STDERR "\n";

	print STDERR "oblique up through negative x axis (expect 1): " .
			signed_crossing_number( -4, -2, -1, 1 ), "\n";
	print STDERR "oblique down through negative x axis (expect -1): " .
			signed_crossing_number( -1, 1, -4, -2 ), "\n";
	print STDERR "\n";

	print STDERR "through origin (expect 2): " . 
			signed_crossing_number( -1, -1, 1, 1 ), "\n";
	print STDERR "through origin other way (expect 2): " . 
			signed_crossing_number( -1, 1, 1, -1 ), "\n";
	print STDERR "\n";

	print STDERR "stops on negative x axis going down (expect -0.5): " .
			signed_crossing_number( -1, 2, -3.3, 0 ), "\n";
	print STDERR "stops on negative x axis going up (expect 0.5): " .
			signed_crossing_number( -1, -2, -3.3, 0 ), "\n";
	print STDERR "starts on negative x axis going down (expect -0.5): " .
			signed_crossing_number( -1, 0, -3.3, -5 ), "\n";
	print STDERR "starts on negative x axis going up (expect 0.5): " .
			signed_crossing_number( -1, 0, -3.3, 4 ), "\n";
	print STDERR "\n";
	
	print STDERR "point in first quadrant (expect 0): " . 
			signed_crossing_number(   1,  1,  1,  1 ), "\n";
	print STDERR "point in second quadrant (expect 0): " . 
			signed_crossing_number( -2,  2, -2,  2 ), "\n";
	print STDERR "point in third quadrant (expect 0): " . 
			signed_crossing_number(  -3, -3, -3, -3 ), "\n";
	print STDERR "point in fourth quadrant (expect 0): " . 
			signed_crossing_number(  4, -4,  4, -4 ), "\n";
	print STDERR "point at origin (expect 2): " . 
			signed_crossing_number( 0, 0, 0, 0 ), "\n";
}

sub test_winding_is_inside_polygon {
	@polygon = ( 1, 1, 1, -1, -1, -1, -1, 1 );
	print STDERR "Inside(expect 1): " .
		is_inside_polygon( 0, 0, @polygon ), "\n";

	@polygon = ( 0, 2, 0, 4, -1, 3, -4, -5 );
	print STDERR "Inside(expect 0): " . 
		is_inside_polygon( 0, 0, @polygon ), "\n";

	@polygon = (  0, 2, 0, 4, -1, 3, -4, -5, 0, 0 );
	print STDERR "Inside(expect 1): " . 
		is_inside_polygon( 0, 0, @polygon ), "\n";

	@polygon = ( -5, 1, -8, 1, -8, -2, -6, 0, -5, -2, );
	print STDERR "Inside(expect 0): " . 
		is_inside_polygon( 0, 0, @polygon ), "\n";

	@polygon = ( -3, -2, -2, 0, 0, 0 );
	print STDERR "Inside(expect 1): " . 
		is_inside_polygon( 0, 0, @polygon ), "\n";

	@polygon = ( 3 );
	print STDERR "Inside(expect blowup): " . 
		is_inside_polygon( 0, 0, @polygon ), "\n";

	@polygon = ( 1, 2, 3, 2, 3, 5, 1, 5, 0, 4 );
	print STDERR "Inside(expect 0): " . 
		is_inside_polygon( -0.001, 4, @polygon ), "\n";
}

# test_winding_signed_crossing_number();
# test_winding_is_inside_polygon();

1;
